package/libnetwork/src/libnetwork.c

```diff
@@ -30,42 +35,151 @@
-  bool debug;
-  curl_socket_t socket_fd;
-} network_transfer_t;
+typedef struct context_node context_node_t;
+
+struct network_context_s {
+
+  network_type_t type;         /**< The type of the network session */
+
+  int fd;                      /**< The input fd to read from */
+  bool debug;                  /**< Set if we are emitted debug information */
+  double gga_xfer_secs;        /**< The number of seconds between GGA upload times */
+
+  time_t last_xfer_time;       /**< The last type GGA was uploaded (for NTRIP only) */
+  curl_socket_t socket_fd;     /**< The socket we're read/writing from/to */
+
+  CURL *curl;                  /**< The current cURL handle */
+
+  curl_off_t bytes_transfered; /**< Current number of bytes transferred */
+  curl_off_t stall_count;      /**< Number of times the bytes transferred have not changed */
+
+  volatile bool shutdown_signaled;
+  volatile bool cycle_connection_signaled;
 
-typedef struct {
-  curl_off_t bytes;
-  curl_off_t count;
-  bool debug;
-} network_progress_t;
+  context_node_t* node;
 
-static volatile bool shutdown_signaled = false;
-static volatile bool cycle_connection_signaled = false;
+  char url[4096];
+};
 
-static time_t last_pipe_warn_time = 0;
+struct context_node {
+  network_context_t context;
+  LIST_ENTRY(context_node) entries;
+};
+
+typedef LIST_HEAD(context_nodes_head, context_node) context_nodes_head_t;
+
+context_nodes_head_t context_nodes_head = LIST_HEAD_INITIALIZER(context_nodes_head);
+bool context_nodes_head_initialized = false;
+
+static network_context_t empty_context = {
+  .type = NETWORK_TYPE_INVALID,
+  .fd = -1,
+  .debug = false,
+  .gga_xfer_secs = -1.0,
+  .last_xfer_time = 0,
+  .socket_fd = CURL_SOCKET_BAD,
+  .curl = NULL,
+  .bytes_transfered = 0,
+  .stall_count = 0,
+  .shutdown_signaled = false,
+  .cycle_connection_signaled = false,
+  .node = NULL,
+  .url = "",
+};
+
+#define NMEA_GGA_FILE "/var/run/nmea_GGA"
 
 void libnetwork_shutdown()
 {
-  shutdown_signaled = true;
+  context_node_t* node;
+  LIST_FOREACH(node, &context_nodes_head, entries) {
+    node->context.shutdown_signaled = true;
+  }
 }
 
 void libnetwork_cycle_connection()
 {
-  cycle_connection_signaled = true;
+  context_node_t* node;
+  LIST_FOREACH(node, &context_nodes_head, entries) {
+    node->context.cycle_connection_signaled = true;
+  }
+}
+
+network_context_t* libnetwork_create(network_type_t type)
+{
+  if (!context_nodes_head_initialized) {
+    LIST_INIT(&context_nodes_head);
+    context_nodes_head_initialized = true;
+  }
+
+  context_node_t* node = malloc(sizeof(context_node_t));
+  if (node == NULL)
+    return NULL;
+
+  memcpy(&node->context, &empty_context, sizeof(empty_context));
+
+  node->context.type = type;
+  node->context.node = node;
+
+  LIST_INSERT_HEAD(&context_nodes_head, node, entries);
+
+  return &node->context;
+}
+
+void libnetwork_destroy(network_context_t **ctx)
+{
+  LIST_REMOVE((*ctx)->node, entries);
+  free(*ctx);
+
+  *ctx = NULL;
```

Comment:
Set `context_nodes_head_initialized = false;` here to be consistent?

Reply:
Inspected the macros, I don't need this since I'm using LIST_HEAD_INITIALIZER (which does the same thing, but at declaration time).